File: Support/Lifecycle/Samples/iOS_UIKitCatalog+Tealium+Lifecycle_ObjC/TealiumHelper.m
//
//  Tracker.m
//  TealiumiOS
//
//  Created by Jason Koo on 11/12/15.
//  Copyright © 2015 Tealium Inc. All rights reserved.
//

#import "TealiumHelper.h"

NSString *const TEALIUM_INSTANCE_ID = @"1";

@implementation TealiumHelper

static TealiumHelper * _sharedInstance;

+ (instancetype) sharedInstance {
    
    if (!_sharedInstance){
        _sharedInstance = [[TealiumHelper alloc] init];
    }
    
    return _sharedInstance;
}

+ (void) startTracking {
    
    TEALConfiguration *configuration = [TEALConfiguration configurationWithAccount:@"tealiummobile"
                                                                           profile:@"main"
                                                                       environment:@"dev"];
    
    [configuration setAutotrackingLifecycleEnabled:YES];
    
    Tealium *tealiumInstance1 = [Tealium newInstanceForKey:TEALIUM_INSTANCE_ID configuration:configuration];
    
    
    [tealiumInstance1 setDelegate:[TealiumHelper sharedInstance]];
    
    [tealiumInstance1 addVolatileDataSources:@{@"volatileTestKey":@"volatileTestValue"}];
    
    if (![tealiumInstance1 respondsToSelector:@selector(launch)]){
        NSLog(@"%s Lifecycle Module not loaded.", __FUNCTION__);
    }
    
}

+ (void) trackEventWithTitle:(NSString *)title dataSources:(NSDictionary *)data {
    
    [[Tealium instanceForKey:TEALIUM_INSTANCE_ID] trackEventWithTitle:title dataSources:data];
}

+ (void) trackViewWithTitle:(NSString *)title dataSources:(NSDictionary *)data {
    
    [[Tealium instanceForKey:TEALIUM_INSTANCE_ID] trackViewWithTitle:title dataSources:data];
}

+ (void) trackType: (TEALDispatchType) eventType title: (NSString *)title dataSources: (NSDictionary *)dataSources {
    
    [[Tealium instanceForKey:TEALIUM_INSTANCE_ID] trackType:eventType title: title dataSources:dataSources completion:nil];
    
}

+ (void) stopTracking{
    
    [Tealium destroyInstanceForKey:TEALIUM_INSTANCE_ID];
    
}

#pragma mark - OPTIONAL TEALIUM DELEGATE

- (BOOL) tealium:(Tealium *)tealium shouldDropDispatch:(TEALDispatch *)dispatch {

    // Add optional tracking suppression logic here - returning YES will destroy
    // any processed dispatch so some conditional must eventually return false
    
    return NO;
    
}

- (BOOL) tealium:(Tealium *)tealium shouldQueueDispatch:(TEALDispatch *)dispatch {

    // Add optional queuing / saving logic here - returning YES will save
    // a dispatch so some condition must eventually return NO.
    
    return NO;
    
}

- (void) tealium:(Tealium *)tealium didQueueDispatch:(TEALDispatch *)dispatch{
    
    // Add optional code here to respond to queuing of dispatches.
    
}

- (void) tealium:(Tealium *)tealium didSendDispatch:(TEALDispatch *)dispatch{
    
    // Add optional code here to respond to sent dispatches.
    
}

- (void) tealium:(Tealium *)tealium webViewIsReady:(id)webView{
    
    // Use this to interact with the Tag Management Dispatcher's webview - available only if Tag Management enabled via remote settings.
}

#pragma mark - EXAMPLE METHODS USING OTHER TEALIUM APIS

+ (void) enableRemoteCommandBlock {

    [[Tealium instanceForKey:TEALIUM_INSTANCE_ID] addRemoteCommandID:@"logger" description:@"An example remote command block" targetQueue:dispatch_get_main_queue() responseBlock:^(TEALRemoteCommandResponse * _Nullable response) {
        
        NSLog(@"%s response: %@", __FUNCTION__, response);
        // Put any code here that can execute on the main thread - ie content
        // modification, A/B testing, etc.
        
    }];
    
    [[Tealium instanceForKey:TEALIUM_INSTANCE_ID] addRemoteCommandID:@"blogger" description:@"An example remote command block" targetQueue:dispatch_get_main_queue() responseBlock:^(TEALRemoteCommandResponse * _Nullable response) {
        
        NSLog(@"%s response: %@", __FUNCTION__, response);
        // Put any code here that can execute on the main thread - ie content
        // modification, A/B testing, etc.
        
    }];
    
    [[Tealium instanceForKey:TEALIUM_INSTANCE_ID] addRemoteCommandID:@"flogger" description:@"An example remote command block" targetQueue:dispatch_get_main_queue() responseBlock:^(TEALRemoteCommandResponse * _Nullable response) {
        
        NSLog(@"%s response: %@", __FUNCTION__, response);
        // Put any code here that can execute on the main thread - ie content
        // modification, A/B testing, etc.
        
    }];
    
    
}


@end

File: Samples/iOS_UIKitCatalog+Tealium_Swift/UIKitCatalog/TealiumHelper.swift
//
//  TealiumHelper.swift
//  Blank+Tealium-Swift
//
//  Created by Jason Koo on 11/16/15.
//  Updated by Christina Sund on 12/08/18
//  Copyright © 2015 Tealium. All rights reserved.
//

/*
*  Using an abstract class like this is the recommended best practice for
*  utilizing analytics or other third party libraries requiring an event
*  trigger with optional data.
*/
import Foundation
import TealiumIOS

class TealiumHelper: NSObject {

    static let shared = TealiumHelper()
    static let tealiumInstanceIdentifier = "1"
    static var traceIdentifier: String?

    class func startTracking() {
        let config = TEALConfiguration.init(account: "tealiummobile", profile: "demo", environment: "dev")
        let tealium = Tealium.newInstance(forKey: tealiumInstanceIdentifier, configuration: config)
        tealium.setDelegate(self.shared)
        TealiumHelper.incrementLifetimeValue(tealium, key: "launches", value: 1)
        guard let volatileDataSources = Tealium.instance(forKey: tealiumInstanceIdentifier)?
            .volatileDataSourcesCopy() else {
            return
        }
        print("Volatile Data: \(volatileDataSources)")
    }

    class func addToVolatileData(key: String, value: Any) {
        Tealium.instance(forKey: tealiumInstanceIdentifier)?.addVolatileDataSources([key: value])
    }

    class func trackEvent(_ title: String, dataSources: [String: AnyObject]) {
        Tealium.instance(forKey: tealiumInstanceIdentifier)?.trackType(.activity,
                                                               title: title,
                                                               dataSources: dataSources,
                                                               completion: nil)
    }

    class func trackView(_ title: String, dataSources: [String: AnyObject]) {
        Tealium.instance(forKey: tealiumInstanceIdentifier)?.trackView(withTitle: title, dataSources: dataSources)
    }

    class func startTrace(_ traceIdentifier: String) {
        if traceIdentifier == self.traceIdentifier {
            return
        }
        self.traceIdentifier = traceIdentifier
        addToVolatileData(key: "tealium_trace_id", value: traceIdentifier)
    }

    class func stopTrace() {
        if nil == traceIdentifier {
            return
        }
        traceIdentifier = nil
        Tealium.instance(forKey: tealiumInstanceIdentifier)?.trackView(withTitle: "killTrace",
                                                               dataSources: ["event": "kill_visitor_session",
                                                                             "cp.trace_id": "XXXXX"])
    }

    class func stopTracking() {
        Tealium.destroyInstance(forKey: tealiumInstanceIdentifier)
    }

}

extension TealiumHelper: TealiumDelegate {

    func tealium(_ tealium: Tealium, shouldDrop dispatch: TEALDispatch) -> Bool {
        // Add optional tracking suppression logic here - returning true will destroy
        // any processed dispatch so some conditional must eventually return false
        return false
    }

    func tealium(_ tealium: Tealium, shouldQueue dispatch: TEALDispatch) -> Bool {
        // Add optional queuing / saving logic here - returning true will save
        // a dispatch so some condition must eventually return false.
        return false
    }

    func tealium(_ tealium: Tealium, didQueue dispatch: TEALDispatch) {
        // Add optional code here to respond to queuing of dispatches.
    }

    func tealium(_ tealium: Tealium, didSend dispatch: TEALDispatch) {
        // Add optional code here to respond to sent dispatches.
    }

    func tealium(_ tealium: Tealium, webViewIsReady webView: Any) {
        // Use this to interact with the Tag Management Dispatcher's webview.
        // This is available only if Tag Management enabled via remote settings.
    }
}

// MARK: Example Methods using other Tealium APIs
extension TealiumHelper {

    class func incrementLifetimeValue(_ tealium: Tealium, key: String, value: Int) {
        var oldValueInStorage:Int? = 0
        let persistentData = tealium.persistentDataSourcesCopy()
        if let persistentDataValue = (persistentData[key]) as? String {
            oldValueInStorage = Int(persistentDataValue)
        }
        let incrementedValue = oldValueInStorage! + value
        let inrementedValueString = NSString(format: "%i", incrementedValue)
        tealium.addPersistentDataSources([key: inrementedValueString])
        print("Current lifetime value for \(key) is \(incrementedValue)")
    }

    class func enableRemoteCommand() {
        Tealium.instance(forKey: tealiumInstanceIdentifier)?
            .addRemoteCommandID("testCommand",
                                description: "An example remote command block",
                                targetQueue: DispatchQueue.main,
                                responseBlock: {(response: TEALRemoteCommandResponse?) -> Void in
                                    print("Remote command response processed - \(String(describing: response))")
                                    // Put any code here that can execute on the main thread - ie content
                                    // modification, A/B testing, etc.
                                })
    }

}
